From d548be05533da2ab1acd8182fe727c8dda66ca0e Mon Sep 17 00:00:00 2001
Nguyễn Long <hoanglong.lumi404@gmail.com>
Date: Thu, 13 Apr 2023 06:23:31 +0000
Subject: [PATCH] HWC2On1Adapter: Fix fence leak

* GL blobs hit this leak, that ultimately results in surfaceflinger to crash and SystemUI soft-resetting every so often when under any form of stressful loads.
* Removes the fence dump from Layer::dump, since:
  a) It was leaking (a dup() without a close())
  b) It's not that useful anyway since it wasn't displaying the actual fence fd

Signed-off-by: Nguyễn Long <hoanglong.lumi404@gmail.com>
---
 graphics/composer/2.1/utils/hwc2on1adapter/HWC2On1Adapter.cpp | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/graphics/composer/2.1/utils/hwc2on1adapter/HWC2On1Adapter.cpp b/graphics/composer/2.1/utils/hwc2on1adapter/HWC2On1Adapter.cpp
index 3d138f7..8347208 100644
--- a/graphics/composer/2.1/utils/hwc2on1adapter/HWC2On1Adapter.cpp
+++ b/graphics/composer/2.1/utils/hwc2on1adapter/HWC2On1Adapter.cpp
@@ -1294,7 +1294,8 @@ Error HWC2On1Adapter::Display::set(hwc_display_contents_1& hwcContents) {
     auto& clientTargetLayer = hwcContents.hwLayers[numLayers - 1];
     if (clientTargetLayer.compositionType == HWC_FRAMEBUFFER_TARGET) {
         clientTargetLayer.handle = mClientTarget.getBuffer();
-        clientTargetLayer.acquireFenceFd = mClientTarget.getFence();
+        close(mClientTarget.getFence());
+        clientTargetLayer.acquireFenceFd = -1;
     } else {
         ALOGE("[%" PRIu64 "] set: last HWC layer wasn't FRAMEBUFFER_TARGET",
                 mId);
-- 
2.40.0

